File: LeetCode/remote/rotate_list.rb
# 
# Submission Details
# 230 / 230 test cases passed.
#   Status: Accepted
#   Runtime: 100 ms
#       
#       Submitted: 0 minutes ago
#
#       https://leetcode.com/submissions/detail/39238024/

class ListNode
     attr_accessor :val, :next
     def initialize(val)
         @val = val
         @next = nil
     end

     def to_s
        return "#{@val} #{@next.nil? ? "" : @next.to_s}"
     end
end

head = nil
5.downto(1).each { |i|
    new = ListNode.new(i)
    new.next = head
    head = new
}

puts head

def rotate_right(head, k)
    return head if head.nil?
    it = head
    count = 0
    while it && it.next
        count += 1
        it = it.next
    end

    count += 1

    prev = it
    it.next = head

    (count - k % count).times { 
        head = head.next
        prev = prev.next
    }

    prev.next = nil
    return head
end

puts rotate_right(head, 2)
puts rotate_right(ListNode.new(1), 4)

File: LeetCode/remote/rotate_list.cs
//  https://leetcode.com/problems/rotate-list/
//  
//  Given a list, rotate the list to the right by k places, where k is non-negative.
//
//  For example:
//  Given 1->2->3->4->5->NULL and k = 2,
//  return 4->5->1->2->3->NULL.

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    //  https://leetcode.com/submissions/detail/51636506/
    //
    //
    //  Submission Details
    //  230 / 230 test cases passed.
    //      Status: Accepted
    //      Runtime: 184 ms
    //          
    //          Submitted: 0 minutes ago
    //
    public ListNode RotateRight(ListNode head, int k) {
        if (head == null || head.next == null || k == 0)
        {
            return head;
        }
        
        var length = 1;
        var it = head;
        while (it.next != null)
        {
            it = it.next;
            length++;
        }
        
        var prev = it;
        it.next = head;
        var times = length - (k % length);
        for (var i = 0; i < times; i++)
        {
            head = head.next;
            prev = prev.next;
        }
        
        prev.next = null;
        return head;
    }
}
